### ISSUE TITLE
[addonservices] wrong implementation of AddonService

### ISSUE THREAD

**Issue by J-N-K** [openhab/openhab-core#2581]
The JavaDoc of the `AddonService` states that  `getAddons(Locale locale)` should be cheap to call and return quickly. The implementation does remote HTTP requests instead of the suggested caching which could take some time. The `refreshSource()` method would be the correct place for updating the cache but is not called before addons are displayed in the UI.

**Comment by J-N-K** [comment]
@cweitkamp The `refreshSource` method is never called. When a request is made to the REST API to get all addons, only `getAddons` is called. So another solution than the one you mentioned in #2633 would be to use a timestamp for the last refresh and then call `refreshSource` after some time.

Please do not do any refactoring to the `AddonService` implementations as that would probably result in conflicts with #2607.

**Comment by cweitkamp** [comment]
I implemented a draft and created a PR against your branch used in #2607.

**Comment by J-N-K** [comment]
I thought about something like that before, but it does not solve the problem, it just mitigates it. Every time the cache is refreshed the `getAddons` method is not returning fast. IMO this needs to be solved at the place where `getAddons` is called:

https://github.com/openhab/openhab-core/blob/cabb3f73150d13f44d647f46af8362e396684b92/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/addons/AddonResource.java#L126-L145

To comply with the documentation, `getAddons` should never refresh but only return the cache. The caller should make sure that `refreshSource` is called when needed. Otherwise the `refreshSource` method as part of the API is not necessary:

https://github.com/openhab/openhab-core/blob/cabb3f73150d13f44d647f46af8362e396684b92/bundles/org.openhab.core/src/main/java/org/openhab/core/addon/AddonService.java#L47-L53

Caching is not needed for `getAddon` (and is also not documented in JavaDoc there), because it doesn't happen very often (only once per add-on, when that add-on is selected in the UI, so it is automatically limited by user-interaction).

**Comment by cweitkamp** [comment]
I agree. There we have to think about the implementation too. Should the `AddonResource` call `refreshSource` first before retrieving all add-ons via `getAddons`?
I don't think that a scheduled polling for `AddonService`s is an option.

The code I submitted to your branch reduces potential load on the target source. This might be considered too as there are already complaints about in the community forum.

**Comment by J-N-K** [comment]
WDYT about [033cf1dc645ae15e780ea30c745991e18d761130?](https://github.com/openhab/openhab-core/pull/2607/commits/033cf1dc645ae15e780ea30c745991e18d761130)?

**Comment by cweitkamp** [comment]
Yes, I like that. It looks much cleaner than my approach.
